AX: Building AXObjectCache::{m_sortedLiveRegionIDs, m_sortedLiveRegionIDs} does `n - 1` unnecessary full-tree walks when processing `n` objects that need sorting at the same time

twilco opened the pull request:
#### 3d2e2734107105b8dba93097ebc61d00ad084868
<pre>
AX: Building AXObjectCache::{m_sortedLiveRegionIDs, m_sortedLiveRegionIDs} does `n - 1` unnecessary full-tree walks when processing `n` objects that need sorting at the same time
<a href="https://bugs.webkit.org/show_bug.cgi?id=295920">https://bugs.webkit.org/show_bug.cgi?id=295920</a>
<a href="https://rdar.apple.com/155817283">rdar://155817283</a>

Reviewed by Joshua Hoffman.

Some webpages will add many live regions all at once. When this happens, we&apos;ll call handleLiveRegionCreated in a loop,
with each iteration calling AXObjectCache::addSortedObject. This function then does what can be a full tree walk to try
to place this single object.

Optimize this in two ways:

  1. Group objects that need sorting, and process them all at once in a new AXObjectCache::addSortedObjects function.

  2. Use a pre-order tree traversal for sorting rather than AXSearchManager. AXSearchManager has a complicated searching
     algorithm that is essentially a pre-order traversal, but does a ton of probably unnecessary work. I&apos;ve tried to
     re-write AXSearchManager to do a simple pre-order traversal, but was unsuccessful at the time. In any case, the simple
     and much faster pre-order traversal is fine for sorting objects, so this commit does that.

This commit also inlines AXIsolatedTree::objectForID by using HashMap::find instead of HashMap::get. HashMap::get wouldn&apos;t
work because it would require including AXIsolatedObject.h in AXIsolatedTree.h, which would cause a circular dependency.
HashMap::find avoids this issue. objectForID is an extremely hot function, so increasing the chance it gets inlined
made a difference in samples taken while debugging the original object-sorting issue.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleMenuOpened):
(WebCore::AXObjectCache::handleLiveRegionCreated):
(WebCore::AXObjectCache::handleRoleChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::findMatchingObjectsInternal):
(WebCore::AXSearchManager::findNextStartingFrom): Deleted.
(WebCore::AXSearchManager::findAllMatchingObjectsIgnoringCache): Deleted.
* Source/WebCore/accessibility/AXSearchManager.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::objectForID const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::objectForID const):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::onDocumentRenderTreeCreation):
(WebCore::AXObjectCache::deferSortForLiveRegion):
(WebCore::AXObjectCache::queueUnsortedObject):
(WebCore::AXObjectCache::platformPerformDeferredCacheUpdate):
(WebCore::AXObjectCache::addSortedObjects):
(WebCore::AXObjectCache::initializeSortedIDLists):
(WebCore::AXObjectCache::addSortedObject): Deleted.

Canonical link: <a href="https://commits.webkit.org/297427@main">https://commits.webkit.org/297427@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c0cb6f78c7d9cf47fa5d45ce48f851e87806506

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111544 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21675 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117587 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61815 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113506 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39792 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84765 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35540 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114491 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25451 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100392 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65206 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24805 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18528 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61407 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94848 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18592 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120878 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38597 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28682 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93695 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39031 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96659 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93520 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23866 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38649 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16398 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34633 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38494 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43959 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38149 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41485 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39851 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -673,7 +673,9 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     AXCoreObject::AccessibilityChildrenVector sortedLiveRegions();
     AXCoreObject::AccessibilityChildrenVector sortedNonRootWebAreas();
-    void addSortedObject(AccessibilityObject&, PreSortedObjectType);
+    void deferSortForLiveRegion(Ref<AccessibilityObject>&&);
```
Could this be deferSortForNewLiveRegion?

minorninth commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -673,7 +673,9 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     AXCoreObject::AccessibilityChildrenVector sortedNonRootWebAreas();
-    void addSortedObject(AccessibilityObject&, PreSortedObjectType);
+    void deferSortForLiveRegion(Ref<AccessibilityObject>&&);
+    void queueUnsortedObject(Ref<AccessibilityObject>&&, PreSortedObjectType);
```
Could this be static?

twilco replied:
I don't think so — the implementation leans on a lot of AXObjectCache state and member functions.

minorninth commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -673,7 +673,9 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
-    void addSortedObject(AccessibilityObject&, PreSortedObjectType);
+    void deferSortForLiveRegion(Ref<AccessibilityObject>&&);
+    void queueUnsortedObject(Ref<AccessibilityObject>&&, PreSortedObjectType);
+    void addSortedObjects(Vector<Ref<AccessibilityObject>>&&, PreSortedObjectType);
```
Could this be static?

twilco replied:
I don't think so — the implementation leans on a lot of AXObjectCache state and member functions.

minorninth commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -943,7 +945,7 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     WeakHashSet<AccessibilityObject> m_deferredUnconnectedObjects;
 #if PLATFORM(MAC)
-    WeakHashSet<Document, WeakPtrImplWithEventTargetData> m_deferredDocumentAddedList;
+    HashMap<PreSortedObjectType, Vector<Ref<AccessibilityObject>>, IntHash<PreSortedObjectType>, WTF::StrongEnumHashTraits<PreSortedObjectType>> m_deferredUnsortedObjects;
```
The previous name m_deferredDocumentAddedList was slightly more clear as to the purpose. Can this be m_deferredUnsortedDocumentedAddedObjects or something like that? Might be nice to preserve the "document added" part.

twilco replied:
We have two types of sorted-object-lists — one for all non-root web areas on the page, and another for the live regions on the page.

The previous data structure only dealt with the former (all non-root web areas on the page), and thus could have the more precise name of `m_deferredDocumentAddedList`. This commit extends the deferral of handling sorting to the live region object type too. So `m_deferredUnsortedDocumentedAddedObjects` wouldn't be accurate for something that also needs to hold deferred live-region sorts too, hence the more general name.

Open to other ideas though, so long as the name encapsulates that both object types can be stored! Could also split the data structure into two separate lists, but that would make AXObjectCache a bit bigger than it has to be?

minorninth commented on Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
```diff
@@ -399,12 +399,33 @@ static void exerciseIsIgnored(AccessibilityObject& object)
+
+void AXObjectCache::deferSortForLiveRegion(Ref<AccessibilityObject>&& object)
+{
+    if (!m_sortedIDListsInitialized)
```
Can you just move the m_sortedIDListsInitialized check into queueUnsortedObject?

twilco replied:
Yes, done it latest commit!

minorninth commented on Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
```diff
@@ -399,12 +399,33 @@ static void exerciseIsIgnored(AccessibilityObject& object)
+    }).iterator;
+    unsortedObjectListIterator->value.appendIfNotContains(WTFMove(object));
+
+    if (!m_performCacheUpdateTimer.isActive() && !m_performingDeferredCacheUpdate)
```
Could we have a schedulePerformCacheUpdateTimerIfNeeded() helper that we could use throughout the code?

twilco replied:
Yes I think that's a good idea! Though maybe for a follow-up PR? As said helper should probably not include the `m_performingDeferredCacheUpdate` check that is used here. This usage relies on the fact that sorting is the last step in `performDeferredCacheUpdate`, which other things cannot rely on.

To explain that a bit more, imagine the helper was as written, including the `m_performingDeferredCacheUpdate` check to avoid starting the timer. Imagine we are mid `performDeferredCacheUpdate`, processing some `m_deferredFoo` list in the middle of the function. And as part of processing that `m_deferredFoo` list, we decide we have to add something to an `m_deferredFoo` list above the one that was already processed (this does happen). Starting the timer is the right thing to do here, but with the `m_performingDeferredCacheUpdate` check, we won't start it.

minorninth reviewed (approved):
Some ideas for readability.

This is a big improvement for now, but I wonder if we should further defer sorting the live regions unless they're actually ever queried, for the case when accessibility is enabled for something other than VoiceOver that doesn't even care about live regions.

twilco committed:
AX: Building AXObjectCache::{m_sortedLiveRegionIDs, m_sortedLiveRegionIDs} does `n - 1` unnecessary full-tree walks when processing `n` objects that need sorting at the same time
https://bugs.webkit.org/show_bug.cgi?id=295920
rdar://155817283

Reviewed by Joshua Hoffman.

Some webpages will add many live regions all at once. When this happens, we'll call handleLiveRegionCreated in a loop,
with each iteration calling AXObjectCache::addSortedObject. This function then does what can be a full tree walk to try
to place this single object.

Optimize this in two ways:

  1. Group objects that need sorting, and process them all at once in a new AXObjectCache::addSortedObjects function.

  2. Use a pre-order tree traversal for sorting rather than AXSearchManager. AXSearchManager has a complicated searching
     algorithm that is essentially a pre-order traversal, but does a ton of probably unnecessary work. I've tried to
     re-write AXSearchManager to do a simple pre-order traversal, but was unsuccessful at the time. In any case, the simple
     and much faster pre-order traversal is fine for sorting objects, so this commit does that.

This commit also inlines AXIsolatedTree::objectForID by using HashMap::find instead of HashMap::get. HashMap::get wouldn't
work because it would require including AXIsolatedObject.h in AXIsolatedTree.h, which would cause a circular dependency.
HashMap::find avoids this issue. objectForID is an extremely hot function, so increasing the chance it gets inlined
made a difference in samples taken while debugging the original object-sorting issue.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleMenuOpened):
(WebCore::AXObjectCache::handleLiveRegionCreated):
(WebCore::AXObjectCache::handleRoleChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::findMatchingObjectsInternal):
(WebCore::AXSearchManager::findNextStartingFrom): Deleted.
(WebCore::AXSearchManager::findAllMatchingObjectsIgnoringCache): Deleted.
* Source/WebCore/accessibility/AXSearchManager.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::objectForID const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::objectForID const):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::onDocumentRenderTreeCreation):
(WebCore::AXObjectCache::deferSortForLiveRegion):
(WebCore::AXObjectCache::queueUnsortedObject):
(WebCore::AXObjectCache::platformPerformDeferredCacheUpdate):
(WebCore::AXObjectCache::addSortedObjects):
(WebCore::AXObjectCache::initializeSortedIDLists):
(WebCore::AXObjectCache::addSortedObject): Deleted.

Canonical link: https://commits.webkit.org/297427@main
